### PR TITLE
[Main] Update Cosmos-SDK to 0.47

### PR DESCRIPTION
**Pull request by colmazia**
Fixed: Update Cosmos-SDK to 0.47

## Implementation details
- update dependency to support Cosmos-sdk 0.47
- Refer to the [upgrading guide](https://github.com/cosmos/cosmos-sdk/blob/release/v0.47.x/UPGRADING.md)
- update oracle params to support Cosmos-sdk 0.47

## Please ensure the following requirements are met before submitting a pull request:

- [x] The pull request is targeted against the correct target branch
- [ ] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [x] The pull request includes a description of the implementation/work done in detail.
- [ ] The pull request includes any and all appropriate unit/integration tests
- [ ] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [x] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

File: app/ante.go [hunk 1]
```diff
@@ -48,14 +50,11 @@ func NewAnteHandler(options HandlerOptions) (sdk.AnteHandler, error) {
 		if options.GlobalFeeSubspace.Name() == "" {
 			return nil, sdkerrors.Wrap(sdkerrors.ErrNotFound, "globalfee param store is required for AnteHandler")
 		}
-		if options.StakingSubspace.Name() == "" {
-			return nil, sdkerrors.Wrap(sdkerrors.ErrNotFound, "staking param store is required for AnteHandler")
-		}
 
 		feeChecker := feechecker.NewFeeChecker(
 			options.OracleKeeper,
 			options.GlobalFeeSubspace,
```
Comment (RogerKSI) [inline app/ante.go:56]: please help update GlobalFee to have its own params

File: cmd/bandd/cmd/root.go [hunk 2]
```diff
@@ -84,27 +82,19 @@ func NewRootCmd() (*cobra.Command, params.EncodingConfig) {
...
+			AddGenesisOracleScriptCmd(band.DefaultNodeHome),
 		),
-		genutilcli.ValidateGenesisCmd(band.ModuleBasics),
-		AddGenesisAccountCmd(band.DefaultNodeHome),
-		AddGenesisDataSourceCmd(band.DefaultNodeHome),
-		AddGenesisOracleScriptCmd(band.DefaultNodeHome),
 		tmcli.NewCompletionCmd(rootCmd, true),
 		// testnetCmd(band.ModuleBasics, banktypes.GenesisBalancesIterator{}),
 		debug.Cmd(),
 		config.Cmd(),
```
Comment (RogerKSI) [inline cmd/bandd/cmd/root.go:94]: please add pruning and snapshot command as well. e.g.

```
		pruning.Cmd(newApp, simapp.DefaultNodeHome),
		snapshot.Cmd(newApp),
```

**Review by RogerKSI — CHANGES_REQUESTED** [review]

File: app/ante.go (hunk 1 above)
Comment (colmazia) [reply app/ante.go:56]: ok https://github.com/bandprotocol/chain/pull/320

File: scripts/open_proposal.sh [hunk 3]
```diff
@@ -0,0 +1,18 @@
+export WALLET_NAME=validator
+export CHAIN_ID="bandchain"
+export DAEMON_NAME=bandd
+export DAEMON_HOME=$HOME/.band
+
+bandd config broadcast-mode block
```
[the remark this answers is not in the document]
Comment (colmazia) [reply scripts/open_proposal.sh:6]: file accidentally uploaded

File: scripts/vote_proposal.sh [hunk 4]
```diff
@@ -0,0 +1,14 @@
+export PROPOSAL_ID=1
+export WALLET_NAME=validator
+export CHAIN_ID="bandchain"
+
+bandd config broadcast-mode block
```
[the remark this answers is not in the document]
Comment (colmazia) [reply scripts/vote_proposal.sh:5]: file accidentally uploaded

File: cmd/bandd/cmd/root.go (hunk 2 above)
Comment (colmazia) [reply cmd/bandd/cmd/root.go:94]: ok

File: Makefile [hunk 5]
```diff
@@ -71,11 +71,15 @@ test:
...
+protoVer=0.11.6
+protoImageName=ghcr.io/cosmos/proto-builder:$(protoVer)
+protoImage=$(DOCKER) run --rm -v $(CURDIR):/workspace --workdir /workspace $(protoImageName)
+
 proto-all: proto-format proto-lint proto-gen
 
 proto-gen:
 	@echo "Generating Protobuf files"
-	$(DOCKER) run --rm -v $(CURDIR):/workspace --workdir /workspace $(PROTO_BUILDER_IMAGE) sh ./scripts/protocgen.sh
+	@$(protoImage) sh ./scripts/protocgen.sh
```
Comment (RogerKSI) [inline Makefile:82]: please help update all proto commands as well. I think we don't need PROTO_FORMATTER_IMAGE anymore?

File: cmd/bandd/cmd/root.go [hunk 6]
```diff
@@ -84,27 +84,21 @@ func NewRootCmd() (*cobra.Command, params.EncodingConfig) {
...
+		pruning.Cmd(newApp, band.DefaultNodeHome),
+		// support oracle file folder
```
Comment (RogerKSI) [inline cmd/bandd/cmd/root.go:97]: is this TODO or already supported?

File: app/app.go [hunk 7]
```diff
@@ -278,17 +310,21 @@ func NewBandApp(
 	)
 
 	// set the BaseApp's parameter store
-	bApp.SetParamStore(
-		app.ParamsKeeper.Subspace(baseapp.Paramspace).WithKeyTable(paramstypes.ConsensusParamsKeyTable()),
+	app.ConsensusParamsKeeper = consensusparamkeeper.NewKeeper(
+		appCodec,
+		keys[consensusparamtypes.StoreKey],
+		// 0.47 TODO: change to tech council address
```
Comment (RogerKSI) [inline app/app.go:316]: please remove these todo.

File: proto/oracle/v1/tx.proto [hunk 8]
```diff
@@ -9,6 +9,10 @@ import "google/protobuf/any.proto";
 import "oracle/v1/oracle.proto";
 import "cosmos/base/v1beta1/coin.proto";
 
+import "cosmos_proto/cosmos.proto";
+import "cosmos/msg/v1/msg.proto";
+import "amino/amino.proto";
+
 // Msg defines the oracle Msg service.
 service Msg {
```
Comment (RogerKSI) [inline proto/oracle/v1/tx.proto:17]: please remove unused import in proto files.

File: Makefile (hunk 5 above)
Comment (colmazia) [reply Makefile:82]: ok

File: cmd/bandd/cmd/root.go (hunk 6 above)
Comment (colmazia) [reply cmd/bandd/cmd/root.go:97]: ok

Comment (colmazia) [reply cmd/bandd/cmd/root.go:97]: I think it is the comment that is not intentionally put there

File: app/app.go (hunk 7 above)
Comment (colmazia) [reply app/app.go:316]: ok

File: Makefile [hunk 9]
```diff
@@ -122,7 +126,7 @@ proto-update-deps:
 ## Importing of tendermint protobuf definitions currently requires the
 ## use of `sed` in order to build properly with cosmos-sdk's proto file layout
 ## (which is the standard Buf.build FILE_LAYOUT)
-## Issue link: https://github.com/tendermint/tendermint/issues/5021
+## Issue link: https://github.com/cometbft/cometbft/issues/5021
```
[the remark this answers is not in the document]
Comment (colmazia) [reply Makefile:129]: ok

File: Makefile [hunk 10]
```diff
@@ -68,88 +68,28 @@ test:
 ###                                Protobuf                                 ###
```
Comment (RogerKSI) [inline Makefile:68]: @colmazia please update changelog file as well.

File: x/oracle/keeper/msg_server.go [hunk 11]
```diff
@@ -289,3 +290,29 @@ func (k msgServer) Activate(goCtx context.Context, msg *types.MsgActivate) (*typ
 	))
 	return &types.MsgActivateResponse{}, nil
 }
+
+func (ms msgServer) UpdateParams(
```
Comment (taobun) [inline x/oracle/keeper/msg_server.go:294]: Why not use k (other method use k)?

File: client/grpc/oracle/proof/block_header_merkle_parts.go [hunk 12]
```diff
@@ -33,7 +33,7 @@ func (bp *BlockHeaderMerkleParts) encodeToEthFormat() BlockHeaderMerklePartsEthe
 
 // GetBlockHeaderMerkleParts converts Tendermint block header struct into BlockHeaderMerkleParts for gas-optimized proof verification.
 func GetBlockHeaderMerkleParts(block *types.Header) BlockHeaderMerkleParts {
-	// based on https://github.com/tendermint/tendermint/blob/master/types/block.go#L448
+	// based on https://github.com/cometbft/cometbft/blob/master/types/block.go#L448
```
Comment (taobun) [inline client/grpc/oracle/proof/block_header_merkle_parts.go:36]: Should we keep the issue link to same repo?

File: x/oracle/keeper/msg_server.go (hunk 11 above)
Comment (colmazia) [reply x/oracle/keeper/msg_server.go:294]: ok

File: Makefile (hunk 10 above)
Comment (colmazia) [reply Makefile:68]: ok

File: app/ante.go [hunk 13]
```diff
@@ -38,24 +39,20 @@ func NewAnteHandler(options HandlerOptions) (sdk.AnteHandler, error) {
 	if options.IBCKeeper == nil {
 		return nil, sdkerrors.Wrap(sdkerrors.ErrLogic, "IBC keeper is required for AnteHandler")
 	}
+	if options.GlobalfeeKeeper == nil {
```
Comment (colmazia) [inline app/ante.go:42]: no check for staking keeper?

File: client/grpc/oracle/proof/util.go [hunk 14]
```diff
@@ -7,11 +7,11 @@ import (
...
 	"time"
 
-	gogotypes "github.com/gogo/protobuf/types"
-	tmbytes "github.com/tendermint/tendermint/libs/bytes"
+	tmbytes "github.com/cometbft/cometbft/libs/bytes"
+	gogotypes "github.com/cosmos/gogoproto/types"
 )
 
-// Copied from https://github.com/tendermint/tendermint/blob/master/types/encoding_helper.go
+// Copied from https://github.com/cometbft/cometbft/blob/master/types/encoding_helper.go
```
Comment (colmazia) [inline client/grpc/oracle/proof/util.go:14]: check this link

File: proto/bandchain/v1/oracle/proof.proto [hunk 15]
```diff
@@ -65,19 +65,19 @@ message QueryRequestCountProofResponse {
 // SingleProofResponse is the data structure for response of single proof
 message SingleProofResponse {
   SingleProof proof           = 1 [(gogoproto.nullable) = false];
-  bytes       evm_proof_bytes = 2 [(gogoproto.casttype) = "github.com/tendermint/tendermint/libs/bytes.HexBytes"];
+  bytes       evm_proof_bytes = 2 [(gogoproto.casttype) = "github.com/cometbft/cometbft/libs/bytes.HexBytes"];
```
Comment (colmazia) [inline proto/bandchain/v1/oracle/proof.proto:68]: recheck this

File: x/globalfee/client/cli/query.go [hunk 16]
```diff
@@ -11,22 +11,22 @@ import (
...
-func GetCmdShowMinimumGasPrices() *cobra.Command {
+func GetQueryCmdParams() *cobra.Command {
 	cmd := &cobra.Command{
-		Use:     "minimum-gas-prices",
-		Short:   "Show minimum gas prices",
-		Long:    "Show all minimum gas prices",
+		Use:     "params",
+		Short:   "Show params",
+		Long:    "Show parameter of globalfee module",
 		Aliases: []string{"min"},
```
Comment (colmazia) [inline x/globalfee/client/cli/query.go:30]: wrong alias

File: app/upgrades/v2_6/upgrades.go [hunk 17]
```diff
@@ -3,10 +3,9 @@ package v2_6
 import (
 	"github.com/bandprotocol/chain/v2/app/keepers"
 	"github.com/bandprotocol/chain/v2/app/upgrades"
-	"github.com/bandprotocol/chain/v2/x/globalfee"
 	globalfeetypes "github.com/bandprotocol/chain/v2/x/globalfee/types"
```
Comment (RogerKSI) [inline app/upgrades/v2_6/upgrades.go:6]: import our pkg at bottom

File: x/globalfee/client/cli/query.go (hunk 16 above)
Comment (RogerKSI) [reply x/globalfee/client/cli/query.go]: removed

File: proto/bandchain/v1/oracle/proof.proto (hunk 15 above)
Comment (RogerKSI) [reply proto/bandchain/v1/oracle/proof.proto:68]: look good.

File: client/grpc/oracle/proof/util.go (hunk 14 above)
Comment (RogerKSI) [reply client/grpc/oracle/proof/util.go]: changed to be tag link.

**Review by RogerKSI — APPROVED** [review]

**Review by satawatnack — APPROVED** [review]

File: x/oracle/keeper/owasm.go [hunk 18]
```diff
@@ -22,10 +21,10 @@ func ConvertToOwasmGas(cosmos uint64) uint64 {
 
 // GetSpanSize return maximum value between MaxReportDataSize and MaxCallDataSize
 func (k Keeper) GetSpanSize(ctx sdk.Context) uint64 {
-	if k.MaxReportDataSize(ctx) > k.MaxCalldataSize(ctx) {
-		return k.MaxReportDataSize(ctx)
+	if k.GetParams(ctx).MaxReportDataSize > k.GetParams(ctx).MaxCalldataSize {
```
Comment (taobun) [inline x/oracle/keeper/owasm.go:24]: Should we get params and store in variable first?

**Review by taobun — CHANGES_REQUESTED** [review]

File: x/oracle/types/expected_keepers.go [hunk 19]
```diff
@@ -47,9 +47,17 @@ type DistrKeeper interface {
 
 // ChannelKeeper defines the expected IBC channel keeper
 type ChannelKeeper interface {
-	GetChannel(ctx sdk.Context, srcPort, srcChan string) (channel channeltypes.Channel, found bool)
+	GetChannel(ctx sdk.Context, srcPort, srcChan string) (channel ibcchanneltypes.Channel, found bool)
 	GetNextSequenceSend(ctx sdk.Context, portID, channelID string) (uint64, bool)
```
Comment (taobun) [inline x/oracle/types/expected_keepers.go:51]: Do we still need these functions?

File: x/oracle/keeper/owasm.go (hunk 18 above)
Comment (colmazia) [reply x/oracle/keeper/owasm.go:24]: ok

**Review by taobun — APPROVED** [review]

**Pull request merged**